etc/testing/travis_build_check_and_stash.sh

```diff
@@ -0,0 +1,85 @@
+# reliable at least.
+#
+# So, try a few times to fetch and identify a valid version of the code. If we
+# time out, give up. If we succeed, stash that code in a docker image which
+# later tests can use to get it back.
+
+cd /home/travis/gopath/src/github.com/pachyderm/pachyderm
+
+if [ "${TRAVIS_PULL_REQUEST}" == "false" ]; then
+    # These shenannigans not needed for release and branch builds, hopefully.
+    exit 0
+fi
+
+echo "Detected PR build, checking for consistency..."
+
+tries=0
+while true; do
+    # Check if HEAD is a merge commit with the commit we want to test as one of
+    # the parents (the other one would be current main branch HEAD)
+    parents=$(git rev-list --parents -n 1 HEAD)
+    if [[ "$parents" == *"$TRAVIS_PULL_REQUEST_SHA"* ]]; then
+        set +x
+        echo "===================================================================================="
+        echo "Great, found the commit we're meant to be testing ($TRAVIS_PULL_REQUEST_SHA)"
+        echo "as one of the parents of the HEAD merge preview commit ($parents)"
+        echo "===================================================================================="
+        set -x
+        break
+    else
+        tries=$((tries+1))
+        if [ "$tries" -gt 60 ]; then
+            echo "Gave up waiting for GitHub to give us the right commit";
+            exit 1
+        fi
+        # The commit we want to test wasn't found yet, maybe the GitHub API is out
+        # of date. Let's wait a second and try to update it, then check again.
+        sleep 1
+        set +x
+        echo "===================================================================================="
+        echo "GitHub didn't give us the commit we're meant to be testing ($TRAVIS_PULL_REQUEST_SHA)"
+        echo "as one of the parents of the HEAD merge preview commit ($parents). Trying again..."
+        echo "===================================================================================="
+        set -x
+        git fetch origin +refs/pull/${TRAVIS_PULL_REQUEST}/merge
+        git checkout -qf FETCH_HEAD
+    fi
+done
+
+# Yay, we got a good git state. Save it! Our tests will need it too, and might
+# not be able to get it back then (or might not get it back with the same
+# commit ID at HEAD, which will break trying to pull docker images tagged with
+# the output of pachctl version).
+
+if [[ ! "$TRAVIS_SECURE_ENV_VARS" == "true" ]]; then
+    echo "Need travis env vars so we can auth to docker hub."
+    exit 1
+fi
+docker login -u pachydermbuildbot -p "${DOCKER_PWD}"
+
+cd /home/travis/gopath/src/github.com/pachyderm
+mkdir -p /tmp/save_git_tarball
+tar cf /tmp/save_git_tarball/pachyderm.tar pachyderm
+cd /tmp/save_git_tarball
+
+cat <<EOT >Dockerfile
+FROM ubuntu:xenial
+COPY pachyderm.tar /
+EOT
+
+docker build -t pachyderm/ci_code_bundle:${TRAVIS_BUILD_NUMBER} .
+docker push pachyderm/ci_code_bundle:${TRAVIS_BUILD_NUMBER}
```

Comment:
This would be a little more work, but I feel like it might be cleaner for us to just setup a dedicated bucket for storing CI build assets. We could then store:

* the pachd and worker images
* `pachctl` so it doesn't have to be re-compiled
* `pachyderm.tar`

Reply:
Images are probably better in a container registry (a local one would be slightly better than docker hub for intra-job asset passing), but I agree about the binaries and the tarball. Ok if we make this a follow-up issue?

Reply:
Sure thing

Reply:
circling back to the images, I feel like they'd be better in a bucket because then we wouldn't pollute our docker hub account with a bunch of releases of `pachyderm/pachd` and `pachyderm/worker` -- though it's not a huge deal. It's straight-forward to serialize images via `docker save` and `docker load`, too.

Reply:
sure, makes sense

Reply:
https://github.com/pachyderm/pachyderm/issues/5144